File: src/daytrace/episode.py
from __future__ import annotations

import re
from collections import Counter
from datetime import timedelta

from daytrace.models import (
    ActivityAnchor,
    ActivityEpisode,
    ActivityLabelCount,
    ActivitySession,
    EpisodeBundle,
    OutcomeSignal,
    SessionBundle,
)


BRIDGE_GAP = timedelta(minutes=5)
ONE_SIDED_GAP = timedelta(minutes=2)
GENERIC_LABELS = {
    "chatgpt",
    "powershell",
    "new tab",
    "inbox",
    "slack",
    "outlook",
    "explorer",
    "unknown",
    "unknown activity",
}
COMMUNICATION_TITLE = re.compile(
    r"(?i)(\(DM\).*Slack|\b(?:Inbox|Calendar).*Outlook|\bMicrosoft Teams$)"
)
ANCHOR_PRIORITY = {"repository": 0, "project": 1, "file": 2, "title": 3, "domain": 4}
OUTCOME_LABEL_CODES = {"named-repository-appeared", "checkout-success", "installer-started"}


class EpisodeInvariantError(RuntimeError):
    pass


def is_generic(value: str) -> bool:
    folded = value.casefold().removesuffix(".exe")
    return folded in GENERIC_LABELS or bool(COMMUNICATION_TITLE.search(value))


def session_anchors(session: ActivitySession) -> tuple[ActivityAnchor, ...]:
    found: dict[tuple[str, str], ActivityAnchor] = {}

    def add(kind: str, value: str | None) -> None:
        if not value:
            return
        normalized = " ".join(value.split())
        key = (kind, normalized.casefold())
        found.setdefault(key, ActivityAnchor(kind, normalized))

    for activity_slice in session.slices:
        for context in activity_slice.contexts:
            add("repository", context.url_path)
            add("project", context.project)
            add("file", context.file)
            if context.title and not is_generic(context.title):
                add("title", context.title)
            if context.url_host and not context.url_path:
                add("domain", context.url_host)
        if activity_slice.title and not is_generic(activity_slice.title):
            add("title", activity_slice.title)
    if not found and not is_generic(session.label):
        add("title", session.label)
    return tuple(
        sorted(
            found.values(),
            key=lambda item: (
                ANCHOR_PRIORITY[item.kind],
                item.value.casefold(),
                item.value,
            ),
        )
    )


def anchors_compatible(
    left: tuple[ActivityAnchor, ...], right: tuple[ActivityAnchor, ...]
) -> bool:
    return bool(
        {(item.kind, item.value.casefold()) for item in left}
        & {(item.kind, item.value.casefold()) for item in right}
    )


def _counts(values: list[str]) -> tuple[ActivityLabelCount, ...]:
    counted = Counter(values)
    return tuple(
        ActivityLabelCount(value, count)
        for value, count in sorted(
            counted.items(), key=lambda item: (-item[1], item[0].casefold(), item[0])
        )
    )


def _signals(sessions: tuple[ActivitySession, ...]) -> tuple[OutcomeSignal, ...]:
    grouped: dict[tuple[str, str], list[str]] = {}
    for session in sessions:
        for signal in session.outcome_signals:
            evidence = grouped.setdefault((signal.code, signal.label), [])
            for evidence_id in signal.evidence_ids:
                if evidence_id not in evidence:
                    evidence.append(evidence_id)
    return tuple(
        OutcomeSignal(code, label, tuple(evidence))
        for (code, label), evidence in grouped.items()
    )


def _episode(index: int, sessions: tuple[ActivitySession, ...]) -> ActivityEpisode:
    anchors = tuple(
        dict.fromkeys(anchor for session in sessions for anchor in session_anchors(session))
    )
    anchors = tuple(
        sorted(
            anchors,
            key=lambda item: (
                ANCHOR_PRIORITY[item.kind],
                item.value.casefold(),
                item.value,
            ),
        )
    )
    signals = _signals(sessions)
    if anchors:
        label = anchors[0].value.removeprefix("/")
    else:
        named_outcome = next(
            (item.label for item in signals if item.code in OUTCOME_LABEL_CODES), None
        )
        label = named_outcome or sessions[0].label
    focused = tuple(
        session.focused_seconds
        for session in sessions
        if session.focused_seconds is not None
    )
    return ActivityEpisode(
        episode_id=f"episode-{index:03d}",
        start=min(item.start for item in sessions),
        end=max(item.end for item in sessions),
        active_seconds=sum(item.active_seconds for item in sessions),
        focused_seconds=sum(focused) if focused else None,
        label=label,
        session_ids=tuple(item.session_id for item in sessions),
        anchors=anchors,
        applications=_counts(
            [
                activity_slice.app
                for session in sessions
                for activity_slice in session.slices
                if activity_slice.app
            ]
        ),
        activity_labels=_counts([item.label for item in sessions]),
        outcome_signals=signals,
        evidence_ids=tuple(
            dict.fromkeys(
                evidence_id
                for session in sessions
                for evidence_id in session.evidence_ids
            )
        ),
    )


def compact_sessions(bundle: SessionBundle) -> EpisodeBundle:
    sessions = tuple(sorted(bundle.sessions, key=lambda item: (item.start, item.end)))
    anchors = tuple(session_anchors(item) for item in sessions)
    strong = [index for index, value in enumerate(anchors) if value]
    groups: list[set[int]] = []
    for session_index in strong:
        if groups:
            previous_index = max(groups[-1])
            if (
                anchors_compatible(anchors[previous_index], anchors[session_index])
                and sessions[session_index].start - sessions[previous_index].end
                <= BRIDGE_GAP
            ):
                groups[-1].update(range(previous_index + 1, session_index + 1))
                continue
        groups.append({session_index})

    assigned = {item for group in groups for item in group}
    for session_index, session in enumerate(sessions):
        if session_index in assigned:
            continue
        candidates: set[int] = set()
        for group_index, group in enumerate(groups):
            first = sessions[min(group)]
            last = sessions[max(group)]
            if last.end <= session.start and session.start - last.end <= ONE_SIDED_GAP:
                candidates.add(group_index)
            if session.end <= first.start and first.start - session.end <= ONE_SIDED_GAP:
                candidates.add(group_index)
        if len(candidates) == 1:
            group_index = candidates.pop()
            groups[group_index].add(session_index)
            assigned.add(session_index)

    unassigned = [index for index in range(len(sessions)) if index not in assigned]
    weak_groups: list[set[int]] = []
    for session_index in unassigned:
        if (
            weak_groups
            and session_index == max(weak_groups[-1]) + 1
            and sessions[session_index].start - sessions[max(weak_groups[-1])].end
            <= BRIDGE_GAP
        ):
            weak_groups[-1].add(session_index)
        else:
            weak_groups.append({session_index})
    groups.extend(weak_groups)
    groups.sort(key=lambda group: (sessions[min(group)].start, min(group)))

    episodes = tuple(
        _episode(
            episode_index,
            tuple(sessions[item] for item in sorted(group)),
        )
        for episode_index, group in enumerate(groups, start=1)
    )
    actual_members = [
        session_id for episode in episodes for session_id in episode.session_ids
    ]
    expected_members = [item.session_id for item in sessions]
    if len(actual_members) != len(set(actual_members)) or sorted(actual_members) != sorted(
        expected_members
    ):
        raise EpisodeInvariantError("episode membership was not conserved")
    if sum(item.active_seconds for item in episodes) != sum(
        item.active_seconds for item in sessions
    ):
        raise EpisodeInvariantError("episode duration was not conserved")
    return EpisodeBundle(
        day=bundle.day,
        timezone_name=bundle.timezone_name,
        focused_seconds=bundle.focused_seconds,
        episodes=episodes,
        sessions=sessions,
        diagnostics=bundle.diagnostics,
    )

File: packages/daytrace-core/src/episode.ts
import { EpisodeInvariantError } from "./models.js";
import type {
  ActivityAnchor,
  ActivityEpisode,
  ActivityLabelCount,
  ActivitySession,
  EpisodeBundle,
  OutcomeSignal,
  SessionBundle,
} from "./models.js";
import { timestampMs } from "./time.js";

const BRIDGE_GAP_SECONDS = 5 * 60;
const ONE_SIDED_GAP_SECONDS = 2 * 60;
const GENERIC_LABELS = new Set([
  "chatgpt", "powershell", "new tab", "inbox", "slack", "outlook",
  "explorer", "unknown", "unknown activity",
]);
const COMMUNICATION_TITLE = /(\(DM\).*Slack|\b(?:Inbox|Calendar).*Outlook|\bMicrosoft Teams$)/i;
const ANCHOR_PRIORITY: Readonly<Record<string, number>> = { repository: 0, project: 1, file: 2, title: 3, domain: 4 };
const OUTCOME_LABEL_CODES = new Set(["named-repository-appeared", "checkout-success", "installer-started"]);

export function isGeneric(value: string): boolean {
  const folded = value.toLocaleLowerCase().replace(/\.exe$/, "");
  return GENERIC_LABELS.has(folded) || COMMUNICATION_TITLE.test(value);
}

function compareAnchors(left: ActivityAnchor, right: ActivityAnchor): number {
  return (ANCHOR_PRIORITY[left.kind] ?? 99) - (ANCHOR_PRIORITY[right.kind] ?? 99)
    || left.value.toLocaleLowerCase().localeCompare(right.value.toLocaleLowerCase())
    || left.value.localeCompare(right.value);
}

export function sessionAnchors(session: ActivitySession): readonly ActivityAnchor[] {
  const found = new Map<string, ActivityAnchor>();
  const add = (kind: string, raw: string | undefined): void => {
    if (raw === undefined || raw.length === 0) return;
    const value = raw.trim().split(/\s+/u).join(" ");
    const key = `${kind}\0${value.toLocaleLowerCase()}`;
    if (!found.has(key)) found.set(key, { kind, value });
  };
  for (const item of session.slices) {
    for (const context of item.contexts) {
      add("repository", context.urlPath);
      add("project", context.project);
      add("file", context.file);
      if (context.title !== undefined && !isGeneric(context.title)) add("title", context.title);
      if (context.urlHost !== undefined && context.urlPath === undefined) add("domain", context.urlHost);
    }
    if (item.title !== undefined && !isGeneric(item.title)) add("title", item.title);
  }
  if (found.size === 0 && !isGeneric(session.label)) add("title", session.label);
  return [...found.values()].sort(compareAnchors);
}

export function anchorsCompatible(left: readonly ActivityAnchor[], right: readonly ActivityAnchor[]): boolean {
  const keys = new Set(left.map((item) => `${item.kind}\0${item.value.toLocaleLowerCase()}`));
  return right.some((item) => keys.has(`${item.kind}\0${item.value.toLocaleLowerCase()}`));
}

function counts(values: readonly string[]): readonly ActivityLabelCount[] {
  const totals = new Map<string, number>();
  for (const value of values) totals.set(value, (totals.get(value) ?? 0) + 1);
  return [...totals.entries()]
    .sort(([leftValue, leftCount], [rightValue, rightCount]) => rightCount - leftCount
      || leftValue.toLocaleLowerCase().localeCompare(rightValue.toLocaleLowerCase())
      || leftValue.localeCompare(rightValue))
    .map(([value, count]) => ({ value, count }));
}

function signals(sessions: readonly ActivitySession[]): readonly OutcomeSignal[] {
  const grouped = new Map<string, { code: string; label: string; evidenceIds: string[] }>();
  for (const session of sessions) {
    for (const signal of session.outcomeSignals) {
      const key = `${signal.code}\0${signal.label}`;
      const current = grouped.get(key) ?? { code: signal.code, label: signal.label, evidenceIds: [] };
      for (const evidenceId of signal.evidenceIds) {
        if (!current.evidenceIds.includes(evidenceId)) current.evidenceIds.push(evidenceId);
      }
      grouped.set(key, current);
    }
  }
  return [...grouped.values()];
}

function episode(index: number, sessions: readonly ActivitySession[]): ActivityEpisode {
  const unique = new Map<string, ActivityAnchor>();
  for (const session of sessions) {
    for (const anchor of sessionAnchors(session)) {
      const key = `${anchor.kind}\0${anchor.value}`;
      if (!unique.has(key)) unique.set(key, anchor);
    }
  }
  const anchors = [...unique.values()].sort(compareAnchors);
  const outcomeSignals = signals(sessions);
  const namedOutcome = outcomeSignals.find((item) => OUTCOME_LABEL_CODES.has(item.code))?.label;
  const label = anchors[0]?.value.replace(/^\//, "") ?? namedOutcome ?? sessions[0]?.label ?? "Unknown activity";
  const focused = sessions.flatMap((item) => item.focusedSeconds === undefined ? [] : [item.focusedSeconds]);
  return {
    episodeId: `episode-${String(index).padStart(3, "0")}`,
    start: sessions.reduce((value, item) => timestampMs(item.start) < timestampMs(value) ? item.start : value, sessions[0]!.start),
    end: sessions.reduce((value, item) => timestampMs(item.end) > timestampMs(value) ? item.end : value, sessions[0]!.end),
    activeSeconds: sessions.reduce((total, item) => total + item.activeSeconds, 0),
    ...(focused.length === 0 ? {} : { focusedSeconds: focused.reduce((total, value) => total + value, 0) }),
    label,
    sessionIds: sessions.map((item) => item.sessionId),
    anchors,
    applications: counts(sessions.flatMap((item) => item.slices.flatMap((slice) => slice.app === undefined ? [] : [slice.app]))),
    activityLabels: counts(sessions.map((item) => item.label)),
    outcomeSignals,
    evidenceIds: [...new Set(sessions.flatMap((item) => item.evidenceIds))],
  };
}

function gapSeconds(start: string, end: string): number {
  return (timestampMs(start) - timestampMs(end)) / 1000;
}

function isClose(left: number, right: number): boolean {
  return Math.abs(left - right) <= Math.max(1e-9, 1e-12 * Math.max(Math.abs(left), Math.abs(right)));
}

export function compactSessions(bundle: SessionBundle): EpisodeBundle {
  const sessions = [...bundle.sessions].sort(
    (left, right) => timestampMs(left.start) - timestampMs(right.start) || timestampMs(left.end) - timestampMs(right.end),
  );
  const anchors = sessions.map(sessionAnchors);
  const strong = anchors.flatMap((value, index) => value.length > 0 ? [index] : []);
  const groups: Set<number>[] = [];
  for (const sessionIndex of strong) {
    const previousGroup = groups.at(-1);
    if (previousGroup !== undefined) {
      const previousIndex = Math.max(...previousGroup);
      if (
        anchorsCompatible(anchors[previousIndex] ?? [], anchors[sessionIndex] ?? [])
        && gapSeconds(sessions[sessionIndex]!.start, sessions[previousIndex]!.end) <= BRIDGE_GAP_SECONDS
      ) {
        for (let index = previousIndex + 1; index <= sessionIndex; index += 1) previousGroup.add(index);
        continue;
      }
    }
    groups.push(new Set([sessionIndex]));
  }

  const assigned = new Set(groups.flatMap((group) => [...group]));
  sessions.forEach((session, sessionIndex) => {
    if (assigned.has(sessionIndex)) return;
    const candidates = new Set<number>();
    groups.forEach((group, groupIndex) => {
      const first = sessions[Math.min(...group)];
      const last = sessions[Math.max(...group)];
      if (first === undefined || last === undefined) return;
      if (timestampMs(last.end) <= timestampMs(session.start) && gapSeconds(session.start, last.end) <= ONE_SIDED_GAP_SECONDS) candidates.add(groupIndex);
      if (timestampMs(session.end) <= timestampMs(first.start) && gapSeconds(first.start, session.end) <= ONE_SIDED_GAP_SECONDS) candidates.add(groupIndex);
    });
    if (candidates.size === 1) {
      const groupIndex = [...candidates][0];
      if (groupIndex !== undefined) groups[groupIndex]?.add(sessionIndex);
      assigned.add(sessionIndex);
    }
  });

  const weakGroups: Set<number>[] = [];
  sessions.forEach((session, sessionIndex) => {
    if (assigned.has(sessionIndex)) return;
    const previousGroup = weakGroups.at(-1);
    const previousIndex = previousGroup === undefined ? undefined : Math.max(...previousGroup);
    if (
      previousGroup !== undefined
      && previousIndex !== undefined
      && sessionIndex === previousIndex + 1
      && gapSeconds(session.start, sessions[previousIndex]!.end) <= BRIDGE_GAP_SECONDS
    ) previousGroup.add(sessionIndex);
    else weakGroups.push(new Set([sessionIndex]));
  });
  groups.push(...weakGroups);
  groups.sort((left, right) => timestampMs(sessions[Math.min(...left)]!.start) - timestampMs(sessions[Math.min(...right)]!.start) || Math.min(...left) - Math.min(...right));

  const episodes = groups.map((group, index) => episode(index + 1, [...group].sort((a, b) => a - b).map((item) => sessions[item]!)));
  const actualMembers = episodes.flatMap((item) => item.sessionIds);
  const expectedMembers = sessions.map((item) => item.sessionId);
  if (new Set(actualMembers).size !== actualMembers.length || [...actualMembers].sort().join("\0") !== [...expectedMembers].sort().join("\0")) {
    throw new EpisodeInvariantError("episode-membership", "episode membership was not conserved");
  }
  const episodeSeconds = episodes.reduce((total, item) => total + item.activeSeconds, 0);
  const sessionSeconds = sessions.reduce((total, item) => total + item.activeSeconds, 0);
  if (!isClose(episodeSeconds, sessionSeconds)) {
    throw new EpisodeInvariantError("episode-duration", "episode duration was not conserved");
  }
  return {
    day: bundle.day,
    timezoneName: bundle.timezoneName,
    ...(bundle.focusedSeconds === undefined ? {} : { focusedSeconds: bundle.focusedSeconds }),
    episodes,
    sessions,
    diagnostics: bundle.diagnostics,
  };
}
